src/crimson/osd/pg.cc

```diff
@@ -1086,41 +1115,10 @@ PG::do_osd_ops(
     },
     // failure_func
     [m, &op_info, obc, this]
-    (const std::error_code& e, const ceph_tid_t& rep_tid, bool record_error) {
+    (const std::error_code& e, const ceph_tid_t& rep_tid) {
     logger().error("do_osd_ops_execute::failure_func {} got error: {} record_error: {}",
-                    *m, e, record_error);
-    epoch_t epoch = get_osdmap_epoch();
-    auto last_complete = peering_state.get_info().last_complete;
-    auto fut = seastar::now();
-    if (record_error && !peering_state.pg_has_reset_since(epoch) && op_info.may_write()) {
-      logger().debug("do_osd_ops_execute::failure_func finding rep_tid {}",
-                      rep_tid);
-      ceph_assert(log_entry_version.contains(rep_tid));
-      auto it = log_entry_update_waiting_on.find(rep_tid);
-      ceph_assert(it != log_entry_update_waiting_on.end());
-      auto it2 = it->second.waiting_on.find(pg_whoami);
-      ceph_assert(it2 != it->second.waiting_on.end());
-      it->second.waiting_on.erase(it2);
-      if (it->second.waiting_on.empty()) {
-        log_entry_update_waiting_on.erase(it);
-        peering_state.complete_write(log_entry_version[rep_tid], last_complete);
-        log_entry_version.erase(rep_tid);
-        logger().debug("do_osd_ops_execute::failure_func write complete,"
-                        " erasing rep_tid {}", rep_tid);
-
-      } else {
-        fut = it->second.all_committed.get_shared_future().then(
-          [this, last_complete, rep_tid] {
-          logger().debug("do_osd_ops_execute::failure_func awaited {}", rep_tid);
-          peering_state.complete_write(log_entry_version[rep_tid], last_complete);
-          ceph_assert(!log_entry_update_waiting_on.contains(rep_tid));
-          return seastar::now();
-        });
-      }
-    }
```

Comment:
Looks this logic/effort is moved from the concurrent wait_repop stage into the exclusive process stage, is it also intentional?

If it isn't intentional, probably leaving it in the concurrent wait_repop phase is better than in the exclusive phase from performance perspective?

Or probably it doesn't matter because it isn't supposed to be frequent, what do you think?

Reply:
I'm inclined to move it into the exclusive stage. This path shouldn't be too frequent and I think that it'll be better to avoid completing the write (by calling `PeeringState::complete_write`)  in the concurrent stage.

Reply:
Updated